Bootstrap script

cmacrae opened the pull request:
Figured I'd spend a little time this evening writing a bash script to bootstrap `nix-darwin`.  
It follows the steps from the README, except it uses `curl` and `unzip` to fetch the repo.  

Straightforward stuff, see what you think @LnL7 🙂

yurrriq commented:
See also:
- https://gist.github.com/LnL7/ed6a23cd803fdb371e853fe970a848a3
- https://github.com/yurrriq/dotfiles/blob/nix/nixbld-setup.sh
- Discussion re: setting up `nix-daemon` on #9

LnL7 commented:
This is awesome!

cmacrae commented:
Cool @yurrriq - you guys reckon this stuff is worth adding to the script, too?  
Perhaps as a prompted question to the user initially, but maybe also as a script option, in case the user initially doesn't set this stuff up, but then later decides they want to?

yurrriq commented:
:+1: to the prompt and to the script option. I think the default should be not to set up `nix-dameon`, so long as it remains [disabled by default](https://github.com/LnL7/nix-darwin/blob/d684bbf2028f066bbeb471bfe9c26ab182348eb6/modules/services/nix-daemon.nix#L17).

yurrriq commented:
I would suggest borrowing heavier from [@Lnl7's script](https://gist.github.com/LnL7/ed6a23cd803fdb371e853fe970a848a3) than mine, as they do basically the same thing but it's more robust and generally in line with your script anyway. I think when I wrote mine I hadn't yet seen it.

cmacrae commented:
@yurrriq Nice one 🙂   
@LnL7 what d'you think? Should I add these functionalities?

LnL7 commented:
Using the daemon currently breaks the nixops bootstrapping if you don't have distributed builds configured. But just creating the users doesn't hurt and makes it easy for people to try the daemon if they want.

LnL7 commented on bootstrap.sh:
```diff
@@ -0,0 +1,133 @@
+    REPO_DOWNLOAD=/tmp/nix-darwin-$(date +%Y%m%d).zip
+    if [ ! -f $REPO_DOWNLOAD ]; then
+	echo -e ""$YELLOW"Fetching nix-darwin repo..."$ESC""
+	curl -Lo $REPO_DOWNLOAD https://github.com/LnL7/nix-darwin/archive/master.zip || \
```
the nix package includes `nix-prefetch-url`, with that it's possible to setup something similar to `nix-channel` for a git repo

cmacrae replied:
Sorry, I'm not sure I follow - I'm pretty new to Nix, so you'll have to explain this one

LnL7 replied:
Yeah, let me make an example.

LnL7 replied:
```
nix-env -p "/nix/var/nix/profiles/per-user/$USER/darwin" --set $(nix-prefetch-url --unpack --print-path --name nix-darwin-17.03 https://github.com/LnL7/nix-darwin/archive/master.tar.gz | grep nix-darwin)
ln -sfn "/nix/var/nix/profiles/per-user/$USER/darwin" "$HOME/.nix-defexpr/darwin"
```

if users have a setup like this I can add a `darwin-channel` command with update and rollback functionality like nix-channel/nix-env

cmacrae replied:
Oh wow, that's really nice! 👌 
I'm making some changes to the script now, so I'll add and try this

LnL7 replied:
I'm not sure why it failed before when I tried it, but nix-channel also works.
```
nix-channel --add https://github.com/LnL7/nix-darwin/archive/master.tar.gz darwin
nix-channel --update
```

LnL7 commented on bootstrap.sh:
```diff
@@ -0,0 +1,133 @@
+
+    echo -e ""$BLUE"You're all done!"$ESC""
+    echo -e "Take a look at "$YELLOW"~/.nixpkgs/darwin-configuration.nix"$ESC" to get started."
+    echo -e "See the README for more information: "$BLUE_UL"https://github.com/LnL7/nix-darwin/blob/master/README.md"$ESC""
```
this should probably also print a warning and some suggestions on how to make sure the environment will get loaded, since `/etc/bashrc` won't be replaced by default

cmacrae replied:
Sure, I'll add some more information :)

yurrriq commented:
:+1: :100:

cmacrae commented:
Okay... so for some reason I'm running into this when bootstrapping now:
```
building the system configuration...
building the system configuration...
writing defaults...
setting up user launchd services...
ls: cannot access '/run/current-system/user/Library/LaunchAgents': No such file or directory
setting up ~/Applications...
warning: ~/Applications is a directory, skipping...
setting up /etc...
setting up launchd services...
ls: cannot access '/run/current-system/Library/LaunchAgents': No such file or directory
ls: cannot access '/run/current-system/Library/LaunchDaemons': No such file or directory
configuring time...
```

Weird, because those directories are most certainly there. This is the end of the output of:
```
export NIX_PATH=darwin=$HOME/.nix-defexpr/darwin:darwin-config=$HOME/.nixpkgs/darwin-configuration.nix:$NIX_PATH
$(nix-build '<darwin>' -A system --no-out-link)/sw/bin/darwin-rebuild build
$(nix-build '<darwin>' -A system --no-out-link)/sw/bin/darwin-rebuild switch
```

Only started happening when I implemented the following in place of `curl` and `unzip`:
```
    nix-env -p "/nix/var/nix/profiles/per-user/$USER/darwin" \
	    --set $(nix-prefetch-url --unpack --print-path \
				     --name nix-darwin-17.03 \
				     https://github.com/LnL7/nix-darwin/archive/master.tar.gz \
			| grep nix-darwin)
    ln -sfn "/nix/var/nix/profiles/per-user/$USER/darwin" ~/.nix-defexpr/darwin
```

As far as I can see, everything looks good (with the repo download and the symlink).  
I've tried cleaning out my install completely several times, just to make sure there's nothing polluting the environment.  

I'll try putting the `curl` & `unzip` method back and see if the problem occurs.

cmacrae commented:
Hmmm... same thing now happening even with the old download method 😒

LnL7 commented:
yeah, sounds like there's a symlink that's not right

cmacrae commented:
Well, I clearly messed something up 😅  
Got back to a working state. So I've added:  
- Nix daemon group/user creation support
- Introduced a flag for the above (in case users want to come back and do so)
- Use `nix-prefetch-url` to grab, unpack, and link the repository - hopefully in support for channels!

LnL7 commented:
@cmacrae Any idea what happened? I might be able to make the activation scripts a bit more robust if I know what went wrong.

cmacrae commented:
@LnL7 Really not sure, I just ran the following again on its own after I was getting this failure:
```
$(nix-build '<darwin>' -A system --no-out-link)/sw/bin/darwin-rebuild build
$(nix-build '<darwin>' -A system --no-out-link)/sw/bin/darwin-rebuild switch
```
And it worked.  

Don't quote me on this, but I vaguely remember (when first trying `nix-darwin` out - before I started the boostrap) getting the same issue when first setting up. I could be mistaken though, if I do remember correctly it was one of those things where I completely disregarded it and simply re-ran my last command.  

Purely conjecture though - I guess if you want any sort of direction, it seems the linking in those scripts for launchd in particular would be where to start. As, in my testing, I'm quite sure I completely eradicated Nix from my system, and all it's symlinks (making use of `mdfind` to ensure there was nothing left over).

LnL7 commented:
Oh, I thought the activation failed halfway through. I know what it is then.

cmacrae commented:
@LnL7 Oh, good - then I'm not going mad!

cmacrae commented:
Okay, it's good to get rid of that output 👍  But it was causing the script to exit, too.  
You can see, it would echo `configuring time...`, then it would just silently exit.  

I can probably reproduce it, if you need some help narrowing this down?

LnL7 commented on bootstrap.sh:
```diff
@@ -0,0 +1,234 @@
+    $(nix-build '<darwin>' -A system --no-out-link)/sw/bin/darwin-rebuild switch
+
+    # Source generated bashrc
+    . /etc/static/bashrc
```
This was causing problems in combination with `set -e`, I disabled completion for the example.

cmacrae replied:
Ah okay, nice one 👍

cmacrae commented:
🎉

LnL7 commented:
Updated the [README](https://github.com/LnL7/nix-darwin#install)

yurrriq commented:
:tada:
